File: current_session/python/39.py
"""
39
"""
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        # sum to target
        # 1 <= candidates <= 200
        # len of candidates <= 30
        # 1 <= target <= 500

        candidates.sort(reverse=True)
        stack = [[(i,c)] for i,c in enumerate(candidates)]
        res = []
        while stack:
            comb = stack.pop()
            s = sum([a[1] for a in comb])
            if s == target:
                res += [a[1] for a in comb],
            elif s < target:
                j = comb[-1][0]
                while target-candidates[j] < 0:
                    j += 1
                while j < len(candidates):
                    stack.append(comb+[(j,candidates[j])])
                    j += 1

        return res

print(Solution().combinationSum([2,3,5],8))
print(Solution().combinationSum([2,3,6,7],7))

File: current_session/c++/39.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

// use backtrack

class Solution {
public:
  vector<vector<int>> res;
  vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
    sort(candidates.begin(), candidates.end());
    helper(candidates, vector<int>{}, 0, target);
    return res;
  }
  void helper(vector<int>& nums, vector<int> tmp, int start, int tgt) {
    if (tgt<0) {
      return;
    } else if (tgt==0) {
      res.push_back(tmp);
    } else {
      for (int i=start; i<nums.size(); ++i) {
        tmp.push_back(nums[i]);
        helper(nums, tmp, i, tgt-nums[i]);
        tmp.pop_back();
      }
    }
  }
};